### PR TITLE
DOC: update documentation for CustomBusinessHour

### LINKED ISSUE #49838: BUG: CustomBusinessHour is to capturing last business hour defined by end parameter
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import datetime as dt
freq = pd.offsets.CustomBusinessHour(weekmask="Sat Sun", start="00:00", end="16:00")
pd.date_range(dt.datetime(2020,1,1), dt.datetime(2020,1,15), freq=freq)
```

### Issue Description

Expected to return every hour of a day from midnight to 16:00, but when including end='16:00' that last hour isn't captured.  Per the doc it states that end defaults to '16:00'.  You can put any hour for end and it doesn't capture that final hour.

```
DatetimeIndex(['2020-01-04 00:00:00', '2020-01-04 01:00:00',
               '2020-01-04 02:00:00', '2020-01-04 03:00:00',
               '2020-01-04 04:00:00', '2020-01-04 05:00:00',
               '2020-01-04 06:00:00', '2020-01-04 07:00:00',
               '2020-01-04 08:00:00', '2020-01-04 09:00:00',
               '2020-01-04 10:00:00', '2020-01-04 11:00:00',
               '2020-01-04 12:00:00', '2020-01-04 13:00:00',
               '2020-01-04 14:00:00', '2020-01-04 15:00:00',
               '2020-01-05 00:00:00', '2020-01-05 01:00:00',
               '2020-01-05 02:00:00', '2020-01-05 03:00:00',
               '2020-01-05 04:00:00', '2020-01-05 05:00:00',
               '2020-01-05 06:00:00', '2020-01-05 07:00:00',
               '2020-01-05 08:00:00', '2020-01-05 09:00:00',
               '2020-01-05 10:00:00', '2020-01-05 11:00:00',
               '2020-01-05 12:00:00', '2020-01-05 13:00:00',
               '2020-01-05 14:00:00', '2020-01-05 15:00:00',
               '2020-01-11 00:00:00', '2020-01-11 01:00:00',
               '2020-01-11 02:00:00', '2020-01-11 03:00:00',
               '2020-01-11 04:00:00', '2020-01-11 05:00:00',
               '2020-01-11 06:00:00', '2020-01-11 07:00:00',
               '2020-01-11 08:00:00', '2020-01-11 09:00:00',
               '2020-01-11 10:00:00', '2020-01-11 11:00:00',
               '2020-01-11 12:00:00', '2020-01-11 13:00:00',
               '2020-01-11 14:00:00', '2020-01-11 15:00:00',
               '2020-01-12 00:00:00', '2020-01-12 01:00:00',
               '2020-01-12 02:00:00', '2020-01-12 03:00:00',
               '2020-01-12 04:00:00', '2020-01-12 05:00:00',
               '2020-01-12 06:00:00', '2020-01-12 07:00:00',
               '2020-01-12 08:00:00', '2020-01-12 09:00:00',
               '2020-01-12 10:00:00', '2020-01-12 11:00:00',
               '2020-01-12 12:00:00', '2020-01-12 13:00:00',
               '2020-01-12 14:00:00', '2020-01-12 15:00:00'],
              dtype='datetime64[ns]', freq='CBH')
```

### Expected Behavior

import pandas as pd
import datetime as dt
freq = pd.offsets.CustomBusinessHour(weekmask="Sat Sun", start="00:00", end="16:00")
pd.date_range(dt.datetime(2020,1,1), dt.datetime(2020,1,15), freq=freq)

Expected:

```
DatetimeIndex(['2020-01-04 00:00:00', '2020-01-04 01:00:00',
               '2020-01-04 02:00:00', '2020-01-04 03:00:00',
               '2020-01-04 04:00:00', '2020-01-04 05:00:00',
               '2020-01-04 06:00:00', '2020-01-04 07:00:00',
               '2020-01-04 08:00:00', '2020-01-04 09:00:00',
               '2020-01-04 10:00:00', '2020-01-04 11:00:00',
               '2020-01-04 12:00:00', '2020-01-04 13:00:00',
               '2020-01-04 14:00:00', '2020-01-04 15:00:00',
               '2020-01-04 16:00:00', '2020-01-05 00:00:00',
               '2020-01-05 01:00:00', '2020-01-05 02:00:00',
               '2020-01-05 03:00:00', '2020-01-05 04:00:00',
               '2020-01-05 05:00:00', '2020-01-05 06:00:00',
               '2020-01-05 07:00:00', '2020-01-05 08:00:00',
               '2020-01-05 09:00:00', '2020-01-05 10:00:00',
               '2020-01-05 11:00:00', '2020-01-05 12:00:00',
               '2020-01-05 13:00:00', '2020-01-05 14:00:00',
               '2020-01-05 15:00:00', '2020-01-05 16:00:00',
               '2020-01-11 00:00:00', '2020-01-11 01:00:00',
               '2020-01-11 02:00:00', '2020-01-11 03:00:00',
               '2020-01-11 04:00:00', '2020-01-11 05:00:00',
               '2020-01-11 06:00:00', '2020-01-11 07:00:00',
               '2020-01-11 08:00:00', '2020-01-11 09:00:00',
               '2020-01-11 10:00:00', '2020-01-11 11:00:00',
               '2020-01-11 12:00:00', '2020-01-11 13:00:00',
               '2020-01-11 14:00:00', '2020-01-11 15:00:00',
               '2020-01-11 16:00:00', '2020-01-12 00:00:00',
               '2020-01-12 01:00:00', '2020-01-12 02:00:00',
               '2020-01-12 03:00:00', '2020-01-12 04:00:00',
               '2020-01-12 05:00:00', '2020-01-12 06:00:00',
               '2020-01-12 07:00:00', '2020-01-12 08:00:00',
               '2020-01-12 09:00:00', '2020-01-12 10:00:00',
               '2020-01-12 11:00:00', '2020-01-12 12:00:00',
               '2020-01-12 13:00:00', '2020-01-12 14:00:00',
               '2020-01-12 15:00:00', '2020-01-12 16:00:00'],
              dtype='datetime64[ns]', freq='CBH')
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 91111fd99898d9dcaa6bf6bedb662db4108da6e6
python           : 3.9.2.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 165 Stepping 5, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.5.1
numpy            : 1.23.4
pytz             : 2022.2.1
dateutil         : 2.8.2
setuptools       : 65.3.0
pip              : 22.2.2
Cython           : None
pytest           : 7.2.0
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 3.0.3
lxml.etree       : 4.9.1
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.4.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : None
brotli           : 
fastparquet      : None
fsspec           : 2022.8.2
gcsfs            : None
matplotlib       : 3.6.1
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : 1.0.10
s3fs             : None
scipy            : 1.8.1
snappy           : None
sqlalchemy       : 1.4.40
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : 2022.2
​

</details>

### PR DESCRIPTION
**Pull request by natmokval**
- [x] closes #49838
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature

Updated documentation for `CustomBusinessHour`. Added missing parameters: `holidays`, `calendar`, and `offset`. Provided more examples.

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -3627,24 +3629,95 @@ cdef class CustomBusinessHour(BusinessHour):
...
+        List/array of dates to exclude from the set of valid business days,
+        passed to ``numpy.busdaycalendar``.
+    calendar : np.busdaycalendar
+        Calendar to integrate.
     start : str, time, or list of str/time, default "09:00"
         Start time of your custom business hour in 24h format.
     end : str, time, or list of str/time, default: "17:00"
         End time of your custom business hour in 24h format.
+    offset : timedelta, default timedelta(0)
+        Time offset to apply.
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:3642]: have you noticed whether this parameter has any effect, or is it unused?

**Review by MarcoGorelli — COMMENTED** [review]
thanks for working on this, it's already much clearer

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -3627,24 +3629,95 @@ cdef class CustomBusinessHour(BusinessHour):
...
+
     >>> ts = pd.Timestamp(2022, 8, 5, 16)
     >>> ts + pd.offsets.CustomBusinessHour()
     Timestamp('2022-08-08 09:00:00')
+
+    We can also change the start and the end of business hours.
+
+    >>> ts = pd.Timestamp(2022, 8, 5, 15)
+    >>> ts + pd.offsets.CustomBusinessHour(start="11:00")
+    Timestamp('2022-08-05 16:00:00')
```
[the remark this answers is not in the document]
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx:3656]: To me this example demonstrates that start time has only effect on the new business day and it complements the next one.
But we can remove this example if you think it’s better.

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -3627,24 +3629,95 @@ cdef class CustomBusinessHour(BusinessHour):
...
+
+    >>> from datetime import time as dt_time
+    >>> ts = pd.Timestamp(2022, 8, 5, 16)
+    >>> ts + pd.offsets.CustomBusinessHour(end=dt_time(19, 0))
     Timestamp('2022-08-05 17:00:00')
-    >>> ts + pd.offsets.CustomBusinessHour(start=[time(9, 0), "20:00"],
-    ...                                    end=["17:00", time(22, 0)])
-    Timestamp('2022-08-05 20:00:00')
+
+    >>> from datetime import time as dt_time
```
[the remark this answers is not in the document]
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx:3667]: Thank you, I’ll do it.

File: pandas/_libs/tslibs/offsets.pyx [hunk 4]
```diff
@@ -3627,24 +3629,95 @@ cdef class CustomBusinessHour(BusinessHour):
...
+    >>> freq = pd.offsets.CustomBusinessHour(start=["06:00", "10:00", "15:00"],
+    ...                                      end=["08:00", "12:00", "17:00"])
+    >>> pd.date_range(dt.datetime(2022, 12, 9), dt.datetime(2022, 12, 13), freq=freq)
+    DatetimeIndex(['2022-12-09 06:00:00', '2022-12-09 07:00:00',
+                   '2022-12-09 10:00:00', '2022-12-09 11:00:00',
+                   '2022-12-09 15:00:00', '2022-12-09 16:00:00',
+                   '2022-12-12 06:00:00', '2022-12-12 07:00:00',
+                   '2022-12-12 10:00:00', '2022-12-12 11:00:00',
+                   '2022-12-12 15:00:00', '2022-12-12 16:00:00'],
+                   dtype='datetime64[ns]', freq='CBH')
```
[the remark this answers is not in the document]
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx:3694]: Thank you :)

File: pandas/_libs/tslibs/offsets.pyx [hunk 5]
```diff
@@ -3627,24 +3629,95 @@ cdef class CustomBusinessHour(BusinessHour):
...
+
+    Business days can be specified by ``weekmask`` parameter.
+
+    >>> import datetime as dt
+    >>> freq = pd.offsets.CustomBusinessHour(weekmask="Mon Wed Fri",
+    ...                                      start="10:00", end="13:00")
+    >>> pd.date_range(dt.datetime(2022, 12, 10), dt.datetime(2022, 12, 18), freq=freq)
```
[the remark this answers is not in the document]
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx:3701]: It’s a great idea. I’ll do it.

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx:3642]: I haven’t checked it yet, but I’ll do it. If the parameter has any effect, I’ll add an example.

Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx]: My checks show that the parameter `offset` hasn't any effect on how the `CustomBusinessHour` works. That is why I don’t mark it as a parameter of `CustomBusinessHour`.

File: pandas/_libs/tslibs/offsets.pyx [hunk 6]
```diff
@@ -3627,24 +3629,91 @@ cdef class CustomBusinessHour(BusinessHour):
...
+    In the example below the business hours start at the default value "09:00"
+    and end at "8:00".
+
+    >>> ts = pd.Timestamp(2022, 8, 5, 22)
+    >>> ts + pd.offsets.CustomBusinessHour(end="8:00")
+    Timestamp('2022-08-05 23:00:00')
+
+    >>> pd.offsets.CustomBusinessHour(end="8:00")
+    <CustomBusinessHour: CBH=09:00-08:00>
+
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:3674]: I'm not really sure what this example demonstrates that's not already shown by the others...would be partial to removing it

**Review by MarcoGorelli — COMMENTED** [review]
Nice! Just one more comment

CI failures are unrelated (and fixed as of https://github.com/pandas-dev/pandas/pull/50205), if you fetch and merge upstream/main then it should be fine

File: pandas/_libs/tslibs/offsets.pyx (hunk 6 above)
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx]: Thank you. I deleted this example and updated my PR.

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me - really minor, but in general it's maybe not necessary to say "In the example below", you can just state what's going to be shown. Anyway, not a blocker, this is fine as-is!

**Comment by natmokval** [comment]
Thank you, MarcoGorelli,  for reviewing my PR. I'll keep your comments in mind.
The cat looks cute, just a little bit sad.

**Pull request merged**